Introduce DataRemovalFrequency for scheduled data removal in the NetworkProcess

sysrqb opened the pull request:
#### 45061230013728ec9c4900b01b12af26dc592b4b
<pre>
Introduce DataRemovalFrequency for scheduled data removal in the NetworkProcess
<a href="https://bugs.webkit.org/show_bug.cgi?id=265598">https://bugs.webkit.org/show_bug.cgi?id=265598</a>
<a href="https://rdar.apple.com/118998314">rdar://118998314</a>

Reviewed by Sihui Liu.

In the current implementation, first-party script-written web site storage is
deleted after 7 operational-days without user interaction on the web site. That
7-day operational period is important for limiting persistent cross-site
tracking, but it has a negative impact on usability by people who visit sites
semi-frequently but longer than 7-operational-days. This change introduces a
lower frequency deletion policy if a web site was always loaded with link
decoration filtering enabled (currently disabled by default).

This patch doesn&apos;t change the current behavior and it is backwards compatible
with the current database. It takes the current boolean value that indicates if
the domain is scheduled for data removal, and it converts that into an integer
that defines the data removal frequency. In the current implementation, 0
indicates that the domain is not scheduled for data removal, 1 indicates that
it is scheduled. This patch redefines the value 1 as indicating the web site
data should be removed on a short frequency (7 operational days), and it
defines the value 2 as indicating the web site data should be removed on a
longer frequency (30 operational days).

More precisely, a domain&apos;s data is deleted in two situations. In the first
case, when the current FirstPartyDataRemovalMode is AllButCookies, website data
is deleted after 7-operational days (without user interaction) regardless of
whether cross-site link decoration was detected. In the second case, the
website&apos;s data is deleted after 7-operational days when it is navigated to with
query parameters from a classified domain. If link decoration filtering is
enabled, then a sites ability to transfer a high-entropy identifier cross-site
is significantly reduced. Therefore, storing web site data for longer periods
of time should be safer.

This change combines the two existing scenarios for removing first-party data
(AllButCookies). When cross-site link decoration is detected and if link
decoration filtering is disabled, then the DataRemovalFrequency remains at
7-operational days. However, if we never load a page with cross-site link
decoration, or if we load a site with link decoration and link decoration
filtering is enabled (such that we filter known-problematic identifiers), then
the DataRemovalFrequency is set to 30-operational days.

* LayoutTests/http/tests/resourceLoadStatistics/aggregate-sorted-data-no-storage-access-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/count-third-party-script-import-in-worker-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/count-third-party-script-loads-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/dont-count-third-party-image-as-third-party-script-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/log-cross-site-load-with-link-decoration-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/log-delayed-client-side-redirects-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/remove-website-data-for-origin-deletes-link-decoration-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/remove-website-data-for-origin-deletes-third-party-script-loads-expected.txt:

Rebaseline

 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion-expected.txt: Added.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion.html: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration.html.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-short-deletion-expected.txt: Added.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-short-deletion.html: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration.html.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-before-short-deletion-expected.txt: Added.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-before-short-deletion.html: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration.html.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-unfiltered-link-decoration-after-short-deletion-expected.txt: Added.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-unfiltered-link-decoration-after-short-deletion.html: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration.html.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-unfiltered-link-decoration-before-short-deletion-expected.txt: Added.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-unfiltered-link-decoration-before-short-deletion.html: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration.html.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt: Added.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration.html: Renamed from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration.html.

New tests.

* LayoutTests/http/tests/storageAccess/aggregate-sorted-data-with-storage-access-expected.txt:
* LayoutTests/http/tests/webAPIStatistics/canvas-read-and-write-data-collection-expected.txt:
* LayoutTests/http/tests/webAPIStatistics/font-load-data-collection-expected.txt:
* LayoutTests/http/tests/webAPIStatistics/screen-functions-accessed-data-collection-expected.txt:

Rebaseline

* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion-expected.txt: Added.
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration-expected.txt: Removed.
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-unfiltered-link-decoration-after-short-deletion-expected.txt: Added.
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt.
* LayoutTests/platform/mac-monterey/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion-expected.txt: Added.
* LayoutTests/platform/mac-monterey/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration-expected.txt: Removed.
* LayoutTests/platform/mac-monterey/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-unfiltered-link-decoration-after-short-deletion-expected.txt: Added.
* LayoutTests/platform/mac-monterey/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt.

These platform exceptions are necessary because they don&apos;t support deleting only script-written cookies while keeping server-written cookies.

* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebKit/NetworkProcess/Classifier/DidFilterKnownLinkDecoration.h: Added.
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::dataRemovalFrequencyToString):
(WebKit::toDataRemovalFrequency):
(WebKit::ResourceLoadStatisticsStore::resetParametersToDefaultValues):
(WebKit::ResourceLoadStatisticsStore::destroyStatements):
(WebKit::ResourceLoadStatisticsStore::merge):
(WebKit::ResourceLoadStatisticsStore::logCrossSiteLoadWithLinkDecoration):
(WebKit::ResourceLoadStatisticsStore::predicateValueForDomain const):
(WebKit::ResourceLoadStatisticsStore::isPrevalentResource const):
(WebKit::ResourceLoadStatisticsStore::isVeryPrevalentResource const):
(WebKit::ResourceLoadStatisticsStore::dataRemovalFrequency const):
(WebKit::ResourceLoadStatisticsStore::setIsScheduledForAllScriptWrittenStorageRemoval):
(WebKit::ResourceLoadStatisticsStore::isGrandfathered const):
(WebKit::ResourceLoadStatisticsStore::domains const):
(WebKit::ResourceLoadStatisticsStore::shouldRemoveAllButCookiesFor):
(WebKit::ResourceLoadStatisticsStore::registrableDomainsToDeleteOrRestrictWebsiteDataFor):
(WebKit::ResourceLoadStatisticsStore::resourceToString const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::logCrossSiteLoadWithLinkDecoration):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::didCommitCrossSiteLoadWithDataTransfer):
(WebKit::NetworkProcess::setCrossSiteLoadWithLinkDecorationForTesting):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
(WKWebsiteDataStoreSetStatisticsCrossSiteLoadWithLinkDecoration):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setStatisticsCrossSiteLoadWithLinkDecorationForTesting:withToHost:withWasFiltered:withCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::didCommitCrossSiteLoadWithDataTransfer):
(WebKit::NetworkProcessProxy::setCrossSiteLoadWithLinkDecorationForTesting):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::setCrossSiteLoadWithLinkDecorationForTesting):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setCrossSiteLoadWithLinkDecorationForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setStatisticsCrossSiteLoadWithLinkDecoration):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::setStatisticsCrossSiteLoadWithLinkDecoration):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::platformSetStatisticsCrossSiteLoadWithLinkDecoration):

Canonical link: <a href="https://commits.webkit.org/274398@main">https://commits.webkit.org/274398@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da613bed4100904d7935bb230ff95d0d9f8340be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38885 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17816 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41230 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41416 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34541 "Build was cancelled. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Compiled WebKit (cancelled)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41192 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20738 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15165 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32577 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39458 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15044 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33718 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13021 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13035 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34637 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42693 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35343 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34973 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38833 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13695 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11293 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37034 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15302 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/33926 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8726 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/14967 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14789 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

johnwilander commented on Source/WebKit/NetworkProcess/NetworkProcess.h:
```diff
@@ -271,7 +271,7 @@ class NetworkProcess : public AuxiliaryProcess, private DownloadManager::Client,
     void setTopFrameUniqueRedirectFrom(PAL::SessionID, TopFrameDomain&&, RedirectedFromDomain&&, CompletionHandler<void()>&&);
     void registrableDomainsWithWebsiteData(PAL::SessionID, OptionSet<WebsiteDataType>, bool shouldNotifyPage, CompletionHandler<void(HashSet<RegistrableDomain>&&)>&&);
-    void didCommitCrossSiteLoadWithDataTransfer(PAL::SessionID, RegistrableDomain&& fromDomain, RegistrableDomain&& toDomain, OptionSet<WebCore::CrossSiteNavigationDataTransfer::Flag>, WebPageProxyIdentifier, WebCore::PageIdentifier);
+    void didCommitCrossSiteLoadWithDataTransfer(PAL::SessionID, RegistrableDomain&& fromDomain, RegistrableDomain&& toDomain, OptionSet<WebCore::CrossSiteNavigationDataTransfer::Flag>, WebPageProxyIdentifier, WebCore::PageIdentifier, bool didFilterLinkDecoration);
```
Please make this a boolean enum DidFilterLinkDecoration { Yes, No } so that it's readable on the caller side too. Unless it for some reason has to be a primitive type. I would also prefer a more precise name such as DidFilterKnownLinkDecoration if it indeed is a case of known keys on a list.

johnwilander commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1258,7 +1259,7 @@ void NetworkProcess::resetCacheMaxAgeCapForPrevalentResources(PAL::SessionID ses
 }
 
-void NetworkProcess::didCommitCrossSiteLoadWithDataTransfer(PAL::SessionID sessionID, RegistrableDomain&& fromDomain, RegistrableDomain&& toDomain, OptionSet<WebCore::CrossSiteNavigationDataTransfer::Flag> navigationDataTransfer, WebPageProxyIdentifier webPageProxyID, WebCore::PageIdentifier webPageID)
+void NetworkProcess::didCommitCrossSiteLoadWithDataTransfer(PAL::SessionID sessionID, RegistrableDomain&& fromDomain, RegistrableDomain&& toDomain, OptionSet<WebCore::CrossSiteNavigationDataTransfer::Flag> navigationDataTransfer, WebPageProxyIdentifier webPageProxyID, WebCore::PageIdentifier webPageID, bool didFilterLinkDecoration)
```
Ditto on a boolean enum.

johnwilander reviewed (commented):
This change log comment needs to be corrected "a domain's data is scheduled for deletion when it is navigated to with query parameters from a classified domain" since deletion is not conditional on query parameters.

See my comments on introducing a boolean enum for clarity.

I don't see where the connection to filtering of known tracking parameters is. Is this patch adding infrastructure that will later be used?

Please add a test case that makes sure that 1) the extension of the time window happens under these conditions, 2) that only script-writeable data is affected by this change, and 3) that all script-writeable data is affected by this change.

szewai commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -298,6 +299,18 @@ bool OperatingDate::operator<=(const OperatingDate& other) const
+static DataRemovalFrequency toDataRemovalFrequency(int value)
+{
+    switch (value) {
+    case 0: return DataRemovalFrequency::None;
```
Maybe "DataRemovalFrequency::Never"?

szewai commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
```diff
@@ -259,7 +261,7 @@ class ResourceLoadStatisticsStore final : public DatabaseUtilities, public CanMa
         bool hadUserInteraction;
         bool grandfathered;
-        bool isScheduledForAllButCookieDataRemoval;
+        DataRemovalFrequency isScheduledForAllButCookieDataRemoval;
```
Can we rename this to dataRemovalFrequency for clarity?

szewai reviewed (commented):
Could you add a test for this?  If link decoration filtering part is not testable, at least we want to ensure entries are removed after 30 days (not removed after 7 days) when they have isScheduledForAllButCookieDataRemoval value as 2 in the database.

szewai commented on LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration_FilterLinkDecorationByDefaultEnabled-expected.txt:
```diff
@@ -0,0 +1,67 @@
+    grandfathered: No
+    TopFrameLinkDecorationsFrom:
+        localhost
+    IsScheduledForAllButCookieDataRemoval: No
```
This should be integer value now. And we might dump the statistics before data deletion to ensure the value is correct.

szewai commented on LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration_FilterLinkDecorationByDefaultEnabled.html:
```diff
@@ -0,0 +1,299 @@
+        checkLocalStorageExists(isAfterDeletion, function () {
+            checkIDBDataStoreExists(isAfterDeletion, () => {
+                if (!isAfterDeletion) {
+                    const advanceTime = !currentAdvancedTime ? daysInSeconds : daysInSeconds*8;
```
Why advancing 8 days each time? It seems we only need to ensure data exists after 7days and not exists after 30days.

szewai commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1936,7 +1953,7 @@ void ResourceLoadStatisticsStore::logFrameNavigation(const RegistrableDomain& ta
 }
 
-void ResourceLoadStatisticsStore::logCrossSiteLoadWithLinkDecoration(const NavigatedFromDomain& fromDomain, const NavigatedToDomain& toDomain)
+void ResourceLoadStatisticsStore::logCrossSiteLoadWithLinkDecoration(const NavigatedFromDomain& fromDomain, const NavigatedToDomain& toDomain, DataRemovalFrequency newDataRemovalFrequency)
```
I think the parameter should be DidFilterKnownLinkDecoration not DataRemovalFrequency, because this would never be DataRemovalFrequency::Never right?

szewai commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -2173,7 +2194,12 @@ bool ResourceLoadStatisticsStore::predicateValueForDomain(SQLiteStatementAutoRes
+
+bool ResourceLoadStatisticsStore::predicateBoolValueForDomain(SQLiteStatementAutoResetScope& predicateStatement, const RegistrableDomain& domain) const
+{
+    return !!predicateValueForDomain(predicateStatement, domain);
```
Might be better to let caller get int value and convert to bool by themselves if needed.

szewai commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1283,23 +1285,22 @@ void NetworkProcess::didCommitCrossSiteLoadWithDataTransfer(PAL::SessionID sessi
         if (auto* session = networkSession(sessionID)) {
             if (auto* resourceLoadStatistics = session->resourceLoadStatistics())
-                resourceLoadStatistics->logCrossSiteLoadWithLinkDecoration(WTFMove(fromDomain), WTFMove(toDomain), [] { });
+                resourceLoadStatistics->logCrossSiteLoadWithLinkDecoration(WTFMove(fromDomain), WTFMove(toDomain), didFilterKnownLinkDecoration == DidFilterKnownLinkDecoration::Yes ? DataRemovalFrequency::Long : DataRemovalFrequency::Short, [] { });
```
`didFilterKnownLinkDecoration` should be passed to `logCrossSiteLoadWithLinkDecoration`, and it will be converted to frequency in `logCrossSiteLoadWithLinkDecoration`

szewai commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -1950,8 +1967,12 @@ void ResourceLoadStatisticsStore::logCrossSiteLoadWithLinkDecoration(const Navig
     insertDomainRelationshipList(topFrameLinkDecorationsFromQuery, HashSet<RegistrableDomain>({ fromDomain }), *toDomainResult.second);
 
+    auto currentDataRemovalFrequency = dataRemovalFrequency(toDomain);
+    if (currentDataRemovalFrequency == DataRemovalFrequency::Short)
```
We should also early return if `currentDataRemovalFrequency == newDataRemovalFrequency`, so next database operations can be skipped.

szewai commented on Source/WebKit/NetworkProcess/Classifier/DidFilterKnownLinkDecoration.h:
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
Nit: 2024

szewai commented on LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt:
```diff
@@ -0,0 +1,47 @@
+After deletion: Client-side persistent cookie does not exist.
+After deletion: Regular server-side cookie exists.
+After deletion: LocalStorage entry does not exist.
+After deletion: IDB entry does not exist.
```
The description says the "website data isn't removed" but the result seems removed. Maybe the description is wrong?

szewai commented on LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration_after_short_deletion-expected.txt:
```diff
@@ -0,0 +1,44 @@
+Before deletion: HttpOnly cookie exists.
+Before deletion: Regular server-side cookie exists.
+Before deletion: LocalStorage entry does exist.
+Before deletion: IDB entry does exist.
```
The description says "website data gets removed" but the result is not removed.
And it seems not right that the log here is "Before deletion" instead of "After deletion".

szewai commented on LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration_after_long_deletion-expected.txt:
```diff
@@ -1,13 +1,22 @@
 Before deletion: HttpOnly cookie exists.
 Before deletion: Regular server-side cookie exists.
 Before deletion: LocalStorage entry does exist.
 Before deletion: IDB entry does exist.
```
Why would there be duplicate "Before deletion" logs? It seems the results should be printed once before deletion, and once after deletion.

sysrqb replied:
Fixed.

szewai commented on LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-unfiltered-link-decoration_before_short_deletion-expected.txt:
```diff
@@ -0,0 +1,44 @@
+Before deletion: HttpOnly cookie exists.
+Before deletion: Regular server-side cookie exists.
+Before deletion: LocalStorage entry does exist.
+Before deletion: IDB entry does exist.
```
Description does not match result, and it seems a bit weird to check data exists "Before deletion". Maybe this test can be merged with the one with 7-day check, and the log can start with "Results:" or "After X operational days:".

sysrqb replied:
I want to check "before deletion" because this allows us to dump and verify the load statistics database state before deletion happens. The current implementation only allows dumping state at the end of the test, and that is why I separated the tests.

szewai reviewed (commented):
The code changes look good to me; some comments about the test.

szewai reviewed (approved):
r=me

sysrqb committed:
Introduce DataRemovalFrequency for scheduled data removal in the NetworkProcess
https://bugs.webkit.org/show_bug.cgi?id=265598
rdar://118998314

Reviewed by Sihui Liu.

In the current implementation, first-party script-written web site storage is
deleted after 7 operational-days without user interaction on the web site. That
7-day operational period is important for limiting persistent cross-site
tracking, but it has a negative impact on usability by people who visit sites
semi-frequently but longer than 7-operational-days. This change introduces a
lower frequency deletion policy if a web site was always loaded with link
decoration filtering enabled (currently disabled by default).

This patch doesn't change the current behavior and it is backwards compatible
with the current database. It takes the current boolean value that indicates if
the domain is scheduled for data removal, and it converts that into an integer
that defines the data removal frequency. In the current implementation, 0
indicates that the domain is not scheduled for data removal, 1 indicates that
it is scheduled. This patch redefines the value 1 as indicating the web site
data should be removed on a short frequency (7 operational days), and it
defines the value 2 as indicating the web site data should be removed on a
longer frequency (30 operational days).

More precisely, a domain's data is deleted in two situations. In the first
case, when the current FirstPartyDataRemovalMode is AllButCookies, website data
is deleted after 7-operational days (without user interaction) regardless of
whether cross-site link decoration was detected. In the second case, the
website's data is deleted after 7-operational days when it is navigated to with
query parameters from a classified domain. If link decoration filtering is
enabled, then a sites ability to transfer a high-entropy identifier cross-site
is significantly reduced. Therefore, storing web site data for longer periods
of time should be safer.

This change combines the two existing scenarios for removing first-party data
(AllButCookies). When cross-site link decoration is detected and if link
decoration filtering is disabled, then the DataRemovalFrequency remains at
7-operational days. However, if we never load a page with cross-site link
decoration, or if we load a site with link decoration and link decoration
filtering is enabled (such that we filter known-problematic identifiers), then
the DataRemovalFrequency is set to 30-operational days.

* LayoutTests/http/tests/resourceLoadStatistics/aggregate-sorted-data-no-storage-access-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/count-third-party-script-import-in-worker-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/count-third-party-script-loads-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/dont-count-third-party-image-as-third-party-script-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/log-cross-site-load-with-link-decoration-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/log-delayed-client-side-redirects-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/remove-website-data-for-origin-deletes-link-decoration-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/remove-website-data-for-origin-deletes-third-party-script-loads-expected.txt:

Rebaseline

 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion-expected.txt: Added.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion.html: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration.html.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-short-deletion-expected.txt: Added.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-short-deletion.html: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration.html.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-before-short-deletion-expected.txt: Added.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-before-short-deletion.html: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration.html.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-unfiltered-link-decoration-after-short-deletion-expected.txt: Added.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-unfiltered-link-decoration-after-short-deletion.html: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration.html.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-unfiltered-link-decoration-before-short-deletion-expected.txt: Added.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-unfiltered-link-decoration-before-short-deletion.html: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration.html.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt: Added.
 * LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration.html: Renamed from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration.html.

New tests.

* LayoutTests/http/tests/storageAccess/aggregate-sorted-data-with-storage-access-expected.txt:
* LayoutTests/http/tests/webAPIStatistics/canvas-read-and-write-data-collection-expected.txt:
* LayoutTests/http/tests/webAPIStatistics/font-load-data-collection-expected.txt:
* LayoutTests/http/tests/webAPIStatistics/screen-functions-accessed-data-collection-expected.txt:

Rebaseline

* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion-expected.txt: Added.
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration-expected.txt: Removed.
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-unfiltered-link-decoration-after-short-deletion-expected.txt: Added.
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt.
* LayoutTests/platform/mac-monterey/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion-expected.txt: Added.
* LayoutTests/platform/mac-monterey/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-link-decoration-expected.txt: Removed.
* LayoutTests/platform/mac-monterey/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-unfiltered-link-decoration-after-short-deletion-expected.txt: Added.
* LayoutTests/platform/mac-monterey/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt.

These platform exceptions are necessary because they don't support deleting only script-written cookies while keeping server-written cookies.

* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebKit/NetworkProcess/Classifier/DidFilterKnownLinkDecoration.h: Added.
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::dataRemovalFrequencyToString):
(WebKit::toDataRemovalFrequency):
(WebKit::ResourceLoadStatisticsStore::resetParametersToDefaultValues):
(WebKit::ResourceLoadStatisticsStore::destroyStatements):
(WebKit::ResourceLoadStatisticsStore::merge):
(WebKit::ResourceLoadStatisticsStore::logCrossSiteLoadWithLinkDecoration):
(WebKit::ResourceLoadStatisticsStore::predicateValueForDomain const):
(WebKit::ResourceLoadStatisticsStore::isPrevalentResource const):
(WebKit::ResourceLoadStatisticsStore::isVeryPrevalentResource const):
(WebKit::ResourceLoadStatisticsStore::dataRemovalFrequency const):
(WebKit::ResourceLoadStatisticsStore::setIsScheduledForAllScriptWrittenStorageRemoval):
(WebKit::ResourceLoadStatisticsStore::isGrandfathered const):
(WebKit::ResourceLoadStatisticsStore::domains const):
(WebKit::ResourceLoadStatisticsStore::shouldRemoveAllButCookiesFor):
(WebKit::ResourceLoadStatisticsStore::registrableDomainsToDeleteOrRestrictWebsiteDataFor):
(WebKit::ResourceLoadStatisticsStore::resourceToString const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::logCrossSiteLoadWithLinkDecoration):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::didCommitCrossSiteLoadWithDataTransfer):
(WebKit::NetworkProcess::setCrossSiteLoadWithLinkDecorationForTesting):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.cpp:
(WKWebsiteDataStoreSetStatisticsCrossSiteLoadWithLinkDecoration):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreRef.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setStatisticsCrossSiteLoadWithLinkDecorationForTesting:withToHost:withWasFiltered:withCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::didCommitCrossSiteLoadWithDataTransfer):
(WebKit::NetworkProcessProxy::setCrossSiteLoadWithLinkDecorationForTesting):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::setCrossSiteLoadWithLinkDecorationForTesting):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setCrossSiteLoadWithLinkDecorationForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setStatisticsCrossSiteLoadWithLinkDecoration):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::setStatisticsCrossSiteLoadWithLinkDecoration):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::platformSetStatisticsCrossSiteLoadWithLinkDecoration):

Canonical link: https://commits.webkit.org/274398@main
